### PR TITLE
Remember #107? well heres that again with less forced cmake hacks

### PR DESCRIPTION
**Pull request by yellows111**
please comment on the gitignore situation with cmake?

### CODE REVIEW

File: src/main/CMakeLists.txt [hunk 1]
```diff
@@ -18,7 +18,16 @@ set_target_properties(collab-vm-server PROPERTIES
...
 collabvm_targetize(collab-vm-server)
 
 # TODO: This binary should be the only target which gets installed,
-# 	sans plugin-headers.
\ No newline at end of file
+# 	sans plugin-headers.
+
+# Dynamic binaries are SO much fun!
+if (WIN32)
+add_custom_command(TARGET collab-vm-server POST_BUILD
```
Comment (modeco80) [inline src/main/CMakeLists.txt]: If you're REALLY gonna do this, 
use https://cmake.org/cmake/help/latest/module/GetPrerequisites.html

This is fragile, only works on MinGW, depends on bash, and depends on scripts I just nuked (because they deserved it.)
Wherehas GetPrerequisites works everywhere (although Windows is the only place we need to copy dlls because windows has to be special)

For vcpkg you'll have to maybe nuke this if that's detected (the Q&D way I will not merge because I'm sure there's some way that's infinitely better would probably involve regex testing CMAKE_TOOLCHAIN_FILE for `/scripts/buildsystems/vcpkg.cmake`) though because vcpkg will copy shared libraries automatically (on any platform). On linux/mac it may fuck up RPATH too so you can just kinda run it there but idk

File: README.md [hunk 2]
```diff
@@ -8,7 +8,7 @@ CollabVM was coded and written by Cosmic Sans, Dartz, Geodude, and modeco80... A
 ### License
 CollabVM Server is licensed under the [GNU Public License 3](https://www.gnu.org/licenses/gpl-3.0.en.html).
 
-The CollabVM Server Plugin Interface "CVMpi" is licensed under the [Lesser GNU Public License 3](https://www.gnu.org/licenses/lgpl-3.0.en.html).
+The CollabVM Server Plugin Interface "CVMpi" is licensed under the [Lesser GNU Public License 3](https://www.gnu.org/licenses/lgpl-3.0.en.html).
```
Comment (modeco80) [inline README.md]: I would reword this because the plugin interface isn't like. completely seperate from the server? the server source comes with and should optionally install sample plugin sources, something sorta like:

```diff
+ CollabVM Server is dual-licensed under the [GNU Public License 3](https://www.gnu.org/licenses/gpl-3.0.en.html) for the server, and the [LGPLv3](https://www.gnu.org/licenses/lgpl-3.0.en.html) license for the plugin interface headers, provided runtime code, and sample plugin code.
```

File: src/main/CMakeLists.txt (hunk 1 above)
Comment (yellows111) [reply src/main/CMakeLists.txt]: yeaaaaaaaaaah i had a feeling this would of been the thing i would of used and would of been bombed for.

using add_custom_command instead of looking into cmake's (what now) more than a thousand functions?

I mean cmake is flexible but it feels too flexible sometimes.

honestly i should of looked around more to see what cmake has support for and now that i know *it exists* i may make a copy_dlls cmake equiv if *possible*. Which it *should* be.

File: README.md (hunk 2 above)
Comment (yellows111) [reply README.md]: i'll look into that.

File: src/main/CMakeLists.txt (hunk 1 above)
Comment (yellows111) [reply src/main/CMakeLists.txt]: >use https://cmake.org/cmake/help/latest/module/GetPrerequisites.html

```
CMake Error at src/main/CMakeLists.txt:29 (get_prerequisites):
  Unknown CMake command "get_prerequisites".
```

> Deprecated since version 3.16: Use file(GET_RUNTIME_DEPENDENCIES) instead.

Deprecated? I think theres an (undocumented) "removed"  missing from this.

also 

>file(GET_RUNTIME_DEPENDENCIES)

yeah i'm not following that deathtrap. (cannot tell difference between 32-bit and 64-bit DLLs, makes it suck. hard.)

Comment (yellows111) [reply src/main/CMakeLists.txt]: the alternative to most of this (except add_custom_command) is `$<TARGET_RUNTIME_DLLS:tgt>`
which has one sliiiiiiiight issue with it
> *New in version 3.21.*

also another problem with it:
is it meant to work? for me with or without list expansion it's empty

Comment (yellows111) [reply src/main/CMakeLists.txt]: Apparently I forgot to check if the "deathtrap" was revised, and as of writing, it was... in 3.21.
I should of probably mentioned this.

File: README.md (hunk 2 above)
Comment (yellows111) [reply README.md]: done https://github.com/yellows111/collab-vm-server/commit/b066e1632614be9ff484e36ce16d10cf9bfc4aee

File: src/main/CMakeLists.txt [hunk 3]
```diff
@@ -8,7 +8,7 @@ add_executable(collab-vm-server
 target_include_directories(collab-vm-server PRIVATE ${CMAKE_CURRENT_SOURCE_DIR})
 target_link_libraries(collab-vm-server PRIVATE ${CMAKE_THREAD_LIBS_INIT} Boost::program_options Boost::filesystem ${CMAKE_DL_LIBS} spdlog::spdlog fmt::fmt collabvm::core collabvm::websocket)
 
-if(WIN32)
+if(WIN32 OR MSYS OR MINGW)
```
Comment (modeco80) [inline src/main/CMakeLists.txt:11]: I mean mingw enviroments would just shortcircuit when WIN32 is true anyways

File: CMakeLists.txt [hunk 4]
```diff
@@ -9,7 +9,14 @@ endif()
 
 include(cmake/Policies.cmake)
 
-project(collab-vm-server)
+# example: 3.0.0-0 = Server 3.0.0, plugin interface/protocol version 0? thought this would be useful
+set(COLLABVM_VERSION_MAJOR 3)
+set(COLLABVM_VERSION_MINOR 0)
+set(COLLABVM_VERSION_PATCH 0)
+set(COLLABVM_PINT_VERSION 1)
```
Comment (modeco80) [inline CMakeLists.txt:16]: Meh, the plugin ABI version is controlled in a .h file and I dunno if I wanna expose that anyways
We will only bump the abi (it's v0 right now) if a huge change is made either way

Also I'm going to be using git tags for versioning probably

Comment (yellows111) [reply CMakeLists.txt:16]: I just wanted to jump on (project VERSION), but we can use them around the cmake files in other ways than just (project VERSION).

File: src/main/CMakeLists.txt (hunk 3 above)
Comment (yellows111) [reply src/main/CMakeLists.txt:11]: makes no sense in that context but apparently some cross-compiling envs specify either MSYS or MINGW and *not* WIN32 for some reason.
